editor/plugins/particle_process_material_editor_plugin.cpp

```diff
@@ -0,0 +1,436 @@
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "particle_process_material_editor_plugin.h"
+
+#include "editor/editor_property_name_processor.h"
+#include "editor/editor_settings.h"
+#include "editor/gui/editor_spin_slider.h"
+#include "scene/gui/base_button.h"
+#include "scene/gui/label.h"
+#include "scene/resources/particle_process_material.h"
+#include "scene/gui/box_container.h"
+
+void ParticleProcessMaterialMinMaxPropertyEditor::_update_sizing() {
+	edit_size = range_edit_widget->get_size();
+	margin = Vector2(range_slider_left_icon->get_width(), (edit_size.y - range_slider_left_icon->get_height()) * 0.5);
+	usable_area = edit_size - margin * 2;
+}
+
+void ParticleProcessMaterialMinMaxPropertyEditor::_range_edit_draw() {
+	ERR_FAIL_COND(range_slider_left_icon.is_null());
+	ERR_FAIL_COND(range_slider_right_icon.is_null());
+	_update_sizing();
+
+	range_edit_widget->draw_rect(Rect2(margin, usable_area), Color(0.3, 0.3, 0.3, 1), false);
+
+	float icon_offset = _get_left_offset() - range_slider_left_icon->get_width() - 1;
+	range_edit_widget->draw_texture(range_slider_left_icon, Vector2(icon_offset, margin.y));
+	icon_offset = _get_right_offset();
+	range_edit_widget->draw_texture(range_slider_right_icon, Vector2(icon_offset, margin.y));
+
+	range_edit_widget->draw_rect(_get_middle_rect(), Color(1, 1, 1, 1));
+}
+
+void ParticleProcessMaterialMinMaxPropertyEditor::_range_edit_gui_input(const Ref<InputEvent> &p_event) {
+	Ref<InputEventMouseButton> mb = p_event;
+	Ref<InputEventMouseMotion> mm = p_event;
+
+	// Prevent unnecessary computations.
+	if ((mb.is_null() || mb->get_button_index() != MouseButton::LEFT) && (mm.is_null() || drag == DRAG_NONE)) {
+		return;
+	}
+
+	ERR_FAIL_COND(range_slider_left_icon.is_null());
+	ERR_FAIL_COND(range_slider_right_icon.is_null());
+	_update_sizing();
+
+	if (mb.is_valid()) {
+		if (mb->is_pressed()) {
+			float left_icon_offset = _get_left_offset() - range_slider_left_icon->get_width() - 1;
+
+			if (Input::get_singleton()->is_key_pressed(Key::SHIFT)) {
```

Comment:
`mb->is_shift_pressed()` here. Speaking of which, this widget could use a tooltip to document this important trick.